In [40]:
# Made By - Shivam Dargan
#Subject  - Computational Statistics
# Group   -  CSBS 1
# Roll No.- 102018021


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
pd.options.mode.chained_assignment = None  # default='warn'

In [58]:
df=pd.read_csv(r'./kc_house_data.csv')
df

id        date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520  10/13/2014  221900.0         3       1.00         1180   
1      6414100192   12/9/2014  538000.0         3       2.25         2570   
2      5631500400   2/25/2015  180000.0         2       1.00          770   
3      2487200875   12/9/2014  604000.0         4       3.00         1960   
4      1954400510   2/18/2015  510000.0         3       2.00         1680   
...           ...         ...       ...       ...        ...          ...   
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
0          5650     1.0           0     0  ...      7        1180   
1          7242     2.0           0     0  ...      7        2170   
2         10000     1.0           0     0  ...      6         770   
3          5000     1.0           0     0  ...      7        1050   
4          8080     1.0           0     0  ...      8        1680   
...         ...     ...         ...   ...  ...    ...         ...   
21592      1131     3.0           0     0  ...      8        1530   
21593      5813     2.0           0     0  ...      8        2310   
21594      1350     2.0           0     0  ...      7        1020   
21595      2388     2.0           0     0  ...      8        1600   
21596      1076     2.0           0     0  ...      7        1020   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                  0      1955             0    98178  47.5112 -122.257   
1                400      1951          1991    98125  47.7210 -122.319   
2                  0      1933             0    98028  47.7379 -122.233   
3                910      1965             0    98136  47.5208 -122.393   
4                  0      1987             0    98074  47.6168 -122.045   
...              ...       ...           ...      ...      ...      ...   
21592              0      2009             0    98103  47.6993 -122.346   
21593              0      2014             0    98146  47.5107 -122.362   
21594              0      2009             0    98144  47.5944 -122.299   
21595              0      2004             0    98027  47.5345 -122.069   
21596              0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[21597 rows x 21 columns]

In [59]:
df=df.iloc[:,[5,2,6]]
df
df.dropna()

sqft_living     price  sqft_lot
0             1180  221900.0      5650
1             2570  538000.0      7242
2              770  180000.0     10000
3             1960  604000.0      5000
4             1680  510000.0      8080
...            ...       ...       ...
21592         1530  360000.0      1131
21593         2310  400000.0      5813
21594         1020  402101.0      1350
21595         1600  400000.0      2388
21596         1020  325000.0      1076

[21597 rows x 3 columns]

# Normalization Of Dataset

In [43]:
max_val = []
min_val = []
for c in df.columns:
    max_val.append(df[c].max())
    min_val.append(df[c].min())
temp = 0
for c in df.columns:
    while(temp < len(df.columns)):
        df[c] = (df[c] - min_val[temp])/(max_val[temp]-min_val[temp])
        temp = temp+1
        break
        
df

sqft_living     price  sqft_lot
0         0.061503  0.018880  0.003108
1         0.167046  0.060352  0.004072
2         0.030372  0.013382  0.005743
3         0.120729  0.069011  0.002714
4         0.099468  0.056678  0.004579
...            ...       ...       ...
21592     0.088079  0.036998  0.000370
21593     0.147304  0.042246  0.003206
21594     0.049355  0.042522  0.000503
21595     0.093394  0.042246  0.001132
21596     0.049355  0.032406  0.000337

[21597 rows x 3 columns]

# Multiple linear Regression model
### price=$B_0$+$B_1$(sqft_living)+$B_2$(sqft_lot)

In [44]:
x1=df.iloc[:,0]   #sqft_living
x2=df.iloc[:,1]   #sqft_lot
y=df.iloc[:,2] #price

## Computing All Of The Parameters

In [51]:
x12=0
x22=0
x1_sum=0   #summation of sqft_living column 
x2_sum=0   #summation of sqft_lot column 
y_sum=0    #summation of y column 
x1_y=0     #summation of multiplication of sqft_living and y column
x2_y=0     #summation of multiplication of sqft_lot and y column
x1_x2=0    #summation of multiplication of sqft_living and sqft_lot column
for i in range(0,df.shape[0]):
    x12+=x1[i]**2
    x22+=x2[i]**2
    x1_sum+=x1[i]
    x2_sum+=x2[i]
    y_sum+=y[i]
    x1_y+=x1[i]*y[i]
    x2_y+=x2[i]*y[i]
    x1_x2+=x1[i]*x2[i]


In [52]:
x1_mean=0    #mean of sqft_living column
x2_mean=0    #mean of sqft_lot column
y_mean=0     #mean of y column
x1_mean=x1_sum/df.shape[0]
x2_mean=x2_sum/df.shape[0]
y_mean=y_sum/df.shape[0]

# Initializing All Parameters

In [53]:
sum_x12=0
sum_x22=0
sum_x1y=0
sum_x2y=0
sum_x1x2=0

## Applying Computational Formulas For MLR

In [54]:
sum_x12=x12-(x1_sum**2)/df.shape[0]
sum_x22=x22-(x2_sum**2)/df.shape[0]
sum_x1y=x1_y-((x1_sum)*(y_sum))/df.shape[0]
sum_x2y=x2_y-((x2_sum)*(y_sum))/df.shape[0]
sum_x1x2=x1_x2-((x1_sum)*(x2_sum))/df.shape[0]

## Calculating $B_1$ & $B_2$ 

In [55]:
b1=((sum_x22)*(sum_x1y)-(sum_x1x2)*(sum_x2y))/((sum_x12)*(sum_x22)-(sum_x1x2)**2)
print("b1 = ",b1)
b2=((sum_x12)*(sum_x2y)-(sum_x1x2)*(sum_x1y))/((sum_x12)*(sum_x22)-(sum_x1x2)**2)
print("b2 = ",b2)

b1 =  0.07828627078296582
b2 =  -0.0326999836307643


# Calculating $B_0$

In [57]:
b0=y_mean-(b1*x1_mean)-(b2*x2_mean)
print("b0 = ",b0)

b0 =  0.0006482188153379377


### Predicting Values Using The Obtained Parameters

In [69]:
print("***********PRICE PREDICTOR******************")
print("Enter Sqft_Living")
sqft_living=float(input())
print("Enter Sqft_Lot")
sqft_lot=float(input())
price = b0 + (b1*sqft_living) + (b2*sqft_lot)
print("The Predicted Price Is: ", price )


***********PRICE PREDICTOR******************
Enter Sqft_Living
0.167
Enter Sqft_Lot
0.004
The Predicted Price Is:  0.013591226101570174


## The Multiple Linear Regression Equation Is : 
### price=$0.0006428$+$0.0782$(sqft_living) - $0.032$(sqft_lot)